internal/mage/util/util.go

```diff
@@ -79,3 +88,118 @@ func DaggerBinary(c *dagger.Client) *dagger.File {
+	sdkHelper      = "dagger-sdk-helper"
+	buildkitRepo   = "github.com/moby/buildkit"
+	buildkitBranch = "v0.10.5"
+)
+
+func DevEngineContainer(c *dagger.Client, arches, oses []string) []*dagger.Container {
+	buildkitRepo := c.Git(buildkitRepo).Branch(buildkitBranch).Tree()
+
+	platformVariants := make([]*dagger.Container, 0, len(arches))
+	for _, arch := range arches {
+		buildkitBase := c.Container(dagger.ContainerOpts{
+			Platform: dagger.Platform("linux/" + arch),
+		}).Build(buildkitRepo)
+		for _, os := range oses {
+			// include each helper for each arch too in case there is a
+			// client/server mismatch
+			for _, arch := range arches {
+				helperBin := GoBase(c).
+					WithEnvVariable("GOOS", os).
+					WithEnvVariable("GOARCH", arch).
+					Exec(dagger.ContainerExecOpts{
+						Args: []string{"go", "build", "-o", "./bin/" + sdkHelper, "-ldflags", "-s -w", "/app/cmd/sdk-helper"},
+					}).
+					File("./bin/" + sdkHelper)
+				buildkitBase = buildkitBase.WithFS(
+					buildkitBase.FS().WithFile("/usr/bin/"+sdkHelper+"-"+os+"-"+arch, helperBin),
+				)
+			}
+		}
+		platformVariants = append(platformVariants, buildkitBase)
+	}
+
+	return platformVariants
+}
+
+func DevEngine(ctx context.Context, c *dagger.Client) (string, error) {
+	tmpfile, err := os.CreateTemp("", "dagger-engine-export")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpfile.Name())
+
+	_, err = c.Container().Export(ctx, tmpfile.Name(), dagger.ContainerExportOpts{
+		PlatformVariants: DevEngineContainer(c, []string{runtime.GOARCH}, []string{runtime.GOOS}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	containerName := "test-dagger-engine"
+	volumeName := "test-dagger-engine"
+	imageName := "localhost/test-dagger-engine:latest"
+
+	// #nosec
+	loadCmd := exec.CommandContext(ctx, "docker", "load", "-i", tmpfile.Name())
+	output, err := loadCmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("docker load failed: %w: %s", err, output)
+	}
+	_, imageID, ok := strings.Cut(string(output), "sha256:")
+	if !ok {
+		return "", fmt.Errorf("unexpected output from docker load: %s", output)
+	}
+	imageID = strings.TrimSpace(imageID)
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"tag",
+		imageID,
+		imageName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker tag: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"rm",
+		"-fv",
+		containerName,
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker rm: %w: %s", err, output)
+	}
+
+	if output, err := exec.CommandContext(ctx, "docker",
+		"run",
+		"-d",
+		"--rm",
+		"-v", volumeName+":/var/lib/buildkit",
+		"--name", containerName,
+		"--privileged",
+		imageName,
+		"--debug",
+	).CombinedOutput(); err != nil {
+		return "", fmt.Errorf("docker run: %w: %s", err, output)
+	}
+
+	return containerName, nil
+}
+
+func WithDevEngine(ctx context.Context, c *dagger.Client, cb func(context.Context, *dagger.Client) error) error {
```

Comment:
I don't have a solution for this, but it's too bad `WithDevEngine` "un-daggerizes" the CI workflow.

That being said, it's the only way

Reply:
Being able to do actual full blown dagger-in-dagger (or docker-in-dagger, etc.) where we do actually nested containers would solve it I think. But that would require enabling privileged execs and setting up mounts very carefully (using the fact that scratch is always a bind mount to avoid overlay-on-overlay, etc.). So yeah, this is the only way *for now* :-)